packages/gatsby-plugin-manifest/src/gatsby-node.js

```diff
@@ -18,48 +18,64 @@ function generateIcons(icons, srcIcon) {
   })
 }
 
-exports.onPostBootstrap = (args, pluginOptions) =>
-  new Promise((resolve, reject) => {
-    const { icon, ...manifest } = pluginOptions
+exports.onPostBootstrap = async ({ createContentDigest }, pluginOptions) => {
+  const { icon, ...manifest } = pluginOptions
 
-    // Delete options we won't pass to the manifest.webmanifest.
+  // Delete options we won't pass to the manifest.webmanifest.
 
-    delete manifest.plugins
-    delete manifest.legacy
-    delete manifest.theme_color_in_head
+  delete manifest.plugins
+  delete manifest.legacy
+  delete manifest.theme_color_in_head
+  delete manifest.cache_busting_mode
 
-    // If icons are not manually defined, use the default icon set.
-    if (!manifest.icons) {
-      manifest.icons = defaultIcons
+  // If icons are not manually defined, use the default icon set.
+  if (!manifest.icons) {
+    manifest.icons = defaultIcons
+  }
+
+  // Determine destination path for icons.
+  const iconPath = path.join(`public`, path.dirname(manifest.icons[0].src))
+
+  //create destination directory if it doesn't exist
+  if (!fs.existsSync(iconPath)) {
+    fs.mkdirSync(iconPath)
+  }
+
+  // Only auto-generate icons if a src icon is defined.
+  if (icon !== undefined) {
+    // Check if the icon exists
+    if (!doesIconExist(icon)) {
+      throw `icon (${icon}) does not exist as defined in gatsby-config.js. Make sure the file exists relative to the root of the site.`
     }
 
-    // Determine destination path for icons.
-    const iconPath = path.join(`public`, path.dirname(manifest.icons[0].src))
+    //add cache busting
+    const cacheMode =
+      typeof pluginOptions.cache_busting_mode !== `undefined`
+        ? pluginOptions.cache_busting_mode
+        : `query`
 
-    //create destination directory if it doesn't exist
-    if (!fs.existsSync(iconPath)) {
-      fs.mkdirSync(iconPath)
+    //if cacheBusting is being done via url query icons must be generated before cachebusting runs
```

Comment:
@moonmeister Here's another part of the cache busting logic that I'm having trouble understanding. Why does the cache mode affect whether the icon is generated before or after the icon's src is set?

Reply:
That shouldn't matter. I think it's just something that was overseen. We just need to write the icons before we generate the digests.

Reply:
@wardpeet @dkthehuman It matters very much. The manifest src paths are modified by the cache busting, 

```js
    if (cacheMode !== `none`) {
      manifest.icons = manifest.icons.map(icon => {
        let newIcon = { ...icon }
        newIcon.src = addDigestToPath(icon.src, iconDigest, cacheMode)
        return newIcon
      })
    }
```

which means if you generated your icons after mutating the icons src for query it'd generate files with the query string (i.e. "icon-256x256.png?c=f34jkl234"). If you were using name based caching you wanted that to happen and need to generate afterwards. 

I'm sure there is a simpler way to handle this...like instead of overwriting the icons array all the time create a copy then overwrite before or after the icon generation based on the type. 

Either-way it is important.

Reply:
thanks @moonmeister, I probably should have read the whole code block instead of that small block :joy:

Reply:
That makes sense. Thanks, guys!